### DISCUSSION TITLE
duckdb 10 times slower than sqlite processing time series

### DISCUSSION THREAD

**Discussion by asuagar** [duckdb/duckdb#2368] [Q&A] [answered]
I am working with maritime vessel positions. There is a global system that recollects the GPS emitted by each maritime vessel.  Each vessel is identified by a number called MMSI. I am working with a two weeks database. Approximately, there are 100k different MMSI and each ships emits  10 times per hour. Therefore, in the database, there are 300M registers.

In the workflow, the time series for each vessel is processed. Therefore, all the registers with a certain a MMSI are selected in a loop. Using duckdb I can only achive 3 it/sec. In [this link](https://colab.research.google.com/drive/1SVcwtve86D19DALJQVX6gSIQzmYW5BRM?usp=sharing), it can be found the code used. I have done the same process using SQLITE. Indexing the MMSI attribute, I achieved more than 30 it/sec.

Is there any way to improve the speed?

**Comment by Mytherin** [comment, marked as answer]
The reason this is slow is because you are doing many point-lookups in a big table (100K point lookups). In SQLite this is faster since they have a big B-tree built on the MMSI attribute, whereas in DuckDB the system is doing 100K full scans. DuckDB does support indexes but is not really built for doing many point lookups in this manner.

A more effective way of doing this would be to obtain all of the timestamps and latitudes/longitudes in one pass, e.g. by running the following `GROUP BY` query:

```sql
SELECT mmsi, LIST(timestamp), LIST(lat), LIST(lon) FROM ais GROUP BY mmsi
```

I'm not sure if LIST types are supported in the Python API yet, if not, then you could add a cast to string here:

```sql
SELECT mmsi, LIST(timestamp)::VARCHAR, LIST(lat)::VARCHAR, LIST(lon)::VARCHAR FROM ais GROUP BY mmsi
```

A nice side effect of this query is that you obtain all of the distinct MMSI's in the same query, so you can replace all of the queries in your example with this one query that should provide you with all the information you need to do the processing.

**Comment by asuagar** [comment]
@Mytherin , thank you for your detailed reply. Before opening the issue, I proved the DuckDB indexes but they did not make a difference, as you stated. As you suggested, I tried both `GROUP BY` queries, but they crashed the Google Colaboratory session. The problem could be in the `LIST()`. However, I tried a `GROUP_CONCAT()` variant and it did the job under 2 min. After that, I processed each MMSI and the performance scaled up to 300 it/sec! This is a huge improvement compared to my first attempt. Thank you so much for giving me that hint.

````
ships = con.execute('''
SELECT 
  mmsi, 
  GROUP_CONCAT(timestamp, ',') AS timestamps, 
  GROUP_CONCAT(lat, ',') AS lats, 
  GROUP_CONCAT(lon, ',') AS lons,
  COUNT(timestamp) AS freq
FROM ais 
GROUP BY mmsi
HAVING freq > 10''').fetchdf()
````

**Reply by bjornasm** [reply]
@Mytherin (or others) - I can't seem to find any reference to `GROUP_CONCAT`in the duckdb documentation, is this deprecated now?

**Reply by Mytherin** [reply]
`group_concat` is an alias to `string_agg`, see [here](https://duckdb.org/docs/sql/aggregates)
